File: src/main/scala/com/vc/medium/L729.scala
package com.vc.medium

class MyCalendar() {

  import java.util.TreeMap
  val map = new TreeMap[Integer, Integer]()

  def book(start: Int, end: Int): Boolean = {
    val lowerKey = map.lowerKey(end)
    if(lowerKey == null || map.get(lowerKey) <= start) {
      map.put(start, end)
      return true
    }
    false
  }

}

/**
  * Your MyCalendar object will be instantiated and called as such:
  * var obj = new MyCalendar()
  * var param_1 = obj.book(start,end)
  */

File: src/main/java/com/vc/medium/L729.java
package com.vc.medium;

import java.util.TreeMap;

class MyCalendar {

    private TreeMap<Integer, Integer> treeMap;

    public MyCalendar() {
        treeMap = new TreeMap<>();
    }

    public boolean book(int start, int end) {
        Integer prev = treeMap.floorKey(start);
        if(prev != null && treeMap.get(prev) > start) return false;

        Integer next = treeMap.lowerKey(end);
        if(next != null && next >= start) return false;

        treeMap.put(start, end);
        return true;
    }
}

/**
 * Your MyCalendar object will be instantiated and called as such:
 * MyCalendar obj = new MyCalendar();
 * boolean param_1 = obj.book(start,end);
 */
